In [1]:
import pandas as pd
import numpy as np
import re
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import classification_report, make_scorer, f1_score, precision_recall_curve
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.multiclass import OneVsRestClassifier

In [2]:
go_file = "/Users/kajolpatel/Desktop/Individual_Project/GOCat/dataset/go-basic.obo"

In [3]:
def parse_obo_file(file_path):
    
    data = []
    current_term = {}
    in_term_block = False
    
    with open(file_path, 'r') as file:
        for line in file:
            line = line.strip()
            if line == '[Term]':  #starting a new term block
                if current_term:
                    data.append(current_term)
                current_term = {}
                in_term_block = True
            elif line == '':
                in_term_block = False  #end of a term block
            elif in_term_block:
                if ': ' in line:
                    key, value = line.split(': ', 1)
                    if key in current_term:  #handling multiple lines of the same key
                        if isinstance(current_term[key], list):
                            current_term[key].append(value)
                        else:
                            current_term[key] = [current_term[key], value]
                    else:
                        current_term[key] = value

    
    if current_term: #add the last term if file does not end with a newline
        data.append(current_term)

    return pd.DataFrame(data)

df = parse_obo_file(go_file)


In [4]:
df = df.rename(columns={'def': 'definition'}) 

In [5]:
df.shape

(47856, 14)

In [6]:
df.head(3)

id                              name           namespace  \
0  GO:0000001         mitochondrion inheritance  biological_process   
1  GO:0000002  mitochondrial genome maintenance  biological_process   
2  GO:0000003             obsolete reproduction  biological_process   

                                          definition  \
0  "The distribution of mitochondria, including t...   
1  "The maintenance of the structure and integrit...   
2  "OBSOLETE. The production of new individuals t...   

                                         synonym  \
0           "mitochondrial inheritance" EXACT []   
1                                            NaN   
2  "reproductive physiological process" EXACT []   

                                                is_a  \
0  [GO:0048308 ! organelle inheritance, GO:004831...   
1            GO:0007005 ! mitochondrion organization   
2                                                NaN   

                     alt_id  \
0                       NaN   
1                       NaN   
2  [GO:0019952, GO:0050876]   

                                             comment is_obsolete replaced_by  \
0                                                NaN         NaN         NaN   
1                                                NaN         NaN         NaN   
2  The reason for obsoletion is that this term is...        true  GO:0022414   

  consider xref subset relationship  
0      NaN  NaN    NaN          NaN  
1      NaN  NaN    NaN          NaN  
2      NaN  NaN    NaN          NaN

In [7]:
df.iloc[0]['is_a']

['GO:0048308 ! organelle inheritance',
 'GO:0048311 ! mitochondrion distribution']

In [8]:
pd.reset_option('display.max_rows')
pd.reset_option('display.max_columns')


In [9]:
print(df[df['is_a'].isna() == False].shape)
print(df[df['is_a'].isna() == True].shape)

(42200, 14)
(5656, 14)


Total records = 47856

42200 records have is_a relationship present

5656 records do not have is_a relationship present

### Data Preprocessing

#### 1. Excluding the records which do not have is_a

In [10]:
df = df[df['is_a'].notna()]

In [11]:
df[df.is_obsolete.isna() == True]

id                                               name  \
0      GO:0000001                          mitochondrion inheritance   
1      GO:0000002                   mitochondrial genome maintenance   
4      GO:0000006  high-affinity zinc transmembrane transporter a...   
5      GO:0000007  low-affinity zinc ion transmembrane transporte...   
7      GO:0000009             alpha-1,6-mannosyltransferase activity   
...           ...                                                ...   
47851  GO:2001313  UDP-4-deoxy-4-formamido-beta-L-arabinopyranose...   
47852  GO:2001314  UDP-4-deoxy-4-formamido-beta-L-arabinopyranose...   
47853  GO:2001315  UDP-4-deoxy-4-formamido-beta-L-arabinopyranose...   
47854  GO:2001316                       kojic acid metabolic process   
47855  GO:2001317                    kojic acid biosynthetic process   

                namespace                                         definition  \
0      biological_process  "The distribution of mitochondria, including t...   
1      biological_process  "The maintenance of the structure and integrit...   
4      molecular_function  "Enables the transfer of zinc ions (Zn2+) from...   
5      molecular_function  "Enables the transfer of a solute or solutes f...   
7      molecular_function  "Catalysis of the transfer of a mannose residu...   
...                   ...                                                ...   
47851  biological_process  "The chemical reactions and pathways involving...   
47852  biological_process  "The chemical reactions and pathways resulting...   
47853  biological_process  "The chemical reactions and pathways resulting...   
47854  biological_process  "The chemical reactions and pathways involving...   
47855  biological_process  "The chemical reactions and pathways resulting...   

                                                 synonym  \
0                   "mitochondrial inheritance" EXACT []   
1                                                    NaN   
4      ["high affinity zinc uptake transmembrane tran...   
5                                                    NaN   
7      "1,6-alpha-mannosyltransferase activity" EXACT []   
...                                                  ...   
47851  "UDP-4-deoxy-4-formamido-beta-L-arabinopyranos...   
47852  ["UDP-4-deoxy-4-formamido-beta-L-arabinopyrano...   
47853  ["UDP-4-deoxy-4-formamido-beta-L-arabinopyrano...   
47854  ["5-hydroxy-2-(hydroxymethyl)-4H-pyran-4-one m...   
47855  ["5-hydroxy-2-(hydroxymethyl)-4H-pyran-4-one a...   

                                                    is_a alt_id comment  \
0      [GO:0048308 ! organelle inheritance, GO:004831...    NaN     NaN   
1                GO:0007005 ! mitochondrion organization    NaN     NaN   
4      GO:0005385 ! zinc ion transmembrane transporte...    NaN     NaN   
5      GO:0005385 ! zinc ion transmembrane transporte...    NaN     NaN   
7              GO:0000030 ! mannosyltransferase activity    NaN     NaN   
...                                                  ...    ...     ...   
47851  [GO:0006040 ! amino sugar metabolic process, G...    NaN     NaN   
47852  [GO:0009227 ! nucleotide-sugar catabolic proce...    NaN     NaN   
47853  [GO:0009226 ! nucleotide-sugar biosynthetic pr...    NaN     NaN   
47854  [GO:0034308 ! primary alcohol metabolic proces...    NaN     NaN   
47855  [GO:0034309 ! primary alcohol biosynthetic pro...    NaN     NaN   

      is_obsolete replaced_by consider  \
0             NaN         NaN      NaN   
1             NaN         NaN      NaN   
4             NaN         NaN      NaN   
5             NaN         NaN      NaN   
7             NaN         NaN      NaN   
...           ...         ...      ...   
47851         NaN         NaN      NaN   
47852         NaN         NaN      NaN   
47853         NaN         NaN      NaN   
47854         NaN         NaN      NaN   
47855         NaN         NaN      NaN   

                                                    xref subse

Checking how many unique values is_a has

In [12]:
exploded_df = df.explode('is_a')

In [13]:
exploded_df['is_a'].value_counts().head(10)

is_a
GO:0110165 ! cellular anatomical entity                                                                                                                                              431
GO:0016616 ! oxidoreductase activity, acting on the CH-OH group of donors, NAD or NADP as acceptor                                                                                   310
GO:0032991 ! protein-containing complex                                                                                                                                              277
GO:0016709 ! oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, NAD(P)H as one donor, and incorporation of one atom of oxygen    261
GO:0016758 ! hexosyltransferase activity                                                                                                                                             208
GO:0048856 ! anatomical structure development                         

#### 15547 unique is_a values

Checking if a GO term can have more than 9 is_a values

In [14]:
df[df['is_a'].apply(lambda x: len(x) > 9 if isinstance(x, list) else False)]

id                                   name           namespace  \
39234  GO:0140872  viridicatumtoxin biosynthetic process  biological_process   

                                              definition  \
39234  "The chemical reactions and pathways resulting...   

                                                 synonym  \
39234  ["viridicatumtoxin anabolism" EXACT [], "virid...   

                                                    is_a alt_id comment  \
39234  [GO:0030639 ! polyketide biosynthetic process,...    NaN     NaN   

      is_obsolete replaced_by consider              xref subset relationship  
39234         NaN         NaN      NaN  MetaCyc:PWY-7659    NaN          NaN

#### 2. Converting the is_a values to only have GO term ids instead of names too :)

In [15]:
def extract_go_terms(s):
    go_terms = []

    if isinstance(s, list):
        for item in s:
            go_terms.extend(re.findall(r'GO:\d{7}', item))
    else:
        go_terms = re.findall(r'GO:\d{7}', s)
    return go_terms if len(go_terms) > 1 else (go_terms[0] if go_terms else None)

In [16]:
df.head(2)

id                              name           namespace  \
0  GO:0000001         mitochondrion inheritance  biological_process   
1  GO:0000002  mitochondrial genome maintenance  biological_process   

                                          definition  \
0  "The distribution of mitochondria, including t...   
1  "The maintenance of the structure and integrit...   

                                synonym  \
0  "mitochondrial inheritance" EXACT []   
1                                   NaN   

                                                is_a alt_id comment  \
0  [GO:0048308 ! organelle inheritance, GO:004831...    NaN     NaN   
1            GO:0007005 ! mitochondrion organization    NaN     NaN   

  is_obsolete replaced_by consider xref subset relationship  
0         NaN         NaN      NaN  NaN    NaN          NaN  
1         NaN         NaN      NaN  NaN    NaN          NaN

In [17]:
df = df[['id','definition','is_a']]

In [18]:
df['is_a'] = df['is_a'].apply(extract_go_terms)

In [19]:
df

id                                         definition  \
0      GO:0000001  "The distribution of mitochondria, including t...   
1      GO:0000002  "The maintenance of the structure and integrit...   
4      GO:0000006  "Enables the transfer of zinc ions (Zn2+) from...   
5      GO:0000007  "Enables the transfer of a solute or solutes f...   
7      GO:0000009  "Catalysis of the transfer of a mannose residu...   
...           ...                                                ...   
47851  GO:2001313  "The chemical reactions and pathways involving...   
47852  GO:2001314  "The chemical reactions and pathways resulting...   
47853  GO:2001315  "The chemical reactions and pathways resulting...   
47854  GO:2001316  "The chemical reactions and pathways involving...   
47855  GO:2001317  "The chemical reactions and pathways resulting...   

                                                   is_a  
0                              [GO:0048308, GO:0048311]  
1                                            GO:0007005  
4                                            GO:0005385  
5                                            GO:0005385  
7                                            GO:0000030  
...                                                 ...  
47851              [GO:0006040, GO:0006793, GO:0009225]  
47852              [GO:0009227, GO:0046348, GO:2001313]  
47853              [GO:0009226, GO:0046349, GO:2001313]  
47854              [GO:0034308, GO:0042180, GO:0120254]  
47855  [GO:0034309, GO:0042181, GO:0120255, GO:2001316]  

[42200 rows x 3 columns]

#### 3. Remove list from definitions

In [20]:
df['definition'] = df['definition'].str.replace(r' \[.*?\]$', '', regex=True)

In [21]:
df['definition'][  0]

'"The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton."'

In [22]:
df['is_a'] = df['is_a'].apply(lambda x: x if isinstance(x, list) else [x])

### Selecting only the records which have is_a as the 10 most frequent values

In [23]:
exploded_df = df.explode('is_a')

In [24]:
exploded_df['is_a'].value_counts().head(10)

is_a
GO:0110165    431
GO:0016616    310
GO:0032991    277
GO:0016709    261
GO:0016758    208
GO:0048856    202
GO:0098797    180
GO:0140513    172
GO:0016747    153
GO:0003006    151
Name: count, dtype: int64

In [29]:
is_a_of_interest = ["GO:0110165","GO:0016616", "GO:0032991", "GO:0016709", "GO:0016758", "GO:0048856", "GO:0098797", "GO:0140513", "GO:0016747", "GO:0003006"]
df['is_a'] = df['is_a'].apply(lambda labels: [label for label in labels if label in is_a_of_interest])
filtered_df = df[df['is_a'].apply(lambda x: any(item in is_a_of_interest for item in (x if isinstance(x, list) else [x])))]

#### Converting the labels to vectors for 10 labels

In [30]:
mlb = MultiLabelBinarizer()
y = mlb.fit_transform(filtered_df['is_a'])
y_df = pd.DataFrame(y, columns=mlb.classes_) 

In [31]:
mlb.classes_

array(['GO:0003006', 'GO:0016616', 'GO:0016709', 'GO:0016747',
       'GO:0016758', 'GO:0032991', 'GO:0048856', 'GO:0098797',
       'GO:0110165', 'GO:0140513'], dtype=object)

In [32]:
y_df

GO:0003006  GO:0016616  GO:0016709  GO:0016747  GO:0016758  GO:0032991  \
0              0           0           0           0           1           0   
1              0           0           0           0           1           0   
2              0           0           0           0           0           0   
3              0           0           0           0           0           0   
4              0           0           0           0           0           0   
...          ...         ...         ...         ...         ...         ...   
2325           0           0           0           0           0           1   
2326           0           0           0           0           0           0   
2327           0           0           0           0           0           1   
2328           0           0           0           0           0           1   
2329           0           0           0           0           0           1   

      GO:0048856  GO:0098797  GO:0110165  GO:0140513  
0              0           0           0           0  
1              0           0           0           0  
2              0           0           0           1  
3              0           0           0           1  
4              0           0           0           1  
...          ...         ...         ...         ...  
2325           0           0           0           0  
2326           0           0           1           0  
2327           0           0           0           0  
2328           0           0           0           0  
2329           0           0           0           0  

[2330 rows x 10 columns]

In [33]:
y_df.columns

Index(['GO:0003006', 'GO:0016616', 'GO:0016709', 'GO:0016747', 'GO:0016758',
       'GO:0032991', 'GO:0048856', 'GO:0098797', 'GO:0110165', 'GO:0140513'],
      dtype='object')

#### Converting definition attribute to feature vectors for 10 labels

Excluding the words which appear in less than 1% of definitions

In [34]:
vectorizer = CountVectorizer(stop_words='english', min_df = 0.01)
X_tfidf = vectorizer.fit_transform(filtered_df['definition'])
X_df = pd.DataFrame(X_tfidf.toarray(), columns=vectorizer.get_feature_names_out())


In [35]:
X_df

11  12  acceptor  acetyl  acid  actin  activation  activity  acyl  \
0      0   0         1       0     0      0           0         0     0   
1      0   0         0       0     0      0           0         0     0   
2      0   0         0       0     0      0           0         0     0   
3      0   0         0       0     0      0           0         1     0   
4      0   0         0       0     0      0           0         0     0   
...   ..  ..       ...     ...   ...    ...         ...       ...   ...   
2325   0   0         0       0     0      0           0         0     0   
2326   0   0         0       0     0      0           0         0     0   
2327   0   0         0       0     0      0           0         0     0   
2328   0   0         0       0     0      0           0         0     0   
2329   0   0         0       0     0      2           0         0     0   

      alpha  ...  time  tissue  transcription  transfer  transmembrane  \
0         0  ...     0       0              0         1              0   
1         0  ...     0       0              0         1              0   
2         0  ...     0       0              0         0              0   
3         0  ...     0       0              0         0              0   
4         0  ...     0       0              1         0              0   
...     ...  ...   ...     ...            ...       ...            ...   
2325      0  ...     0       0              0         0              0   
2326      0  ...     0       0              0         0              0   
2327      0  ...     0       0              0         0              0   
2328      0  ...     0       0              0         0              0   
2329      0  ...     0       0              0         0              0   

      transport  type  udp  wall  yeast  
0             0     0    0     0      0  
1             0     0    0     0      0  
2             0     0    0     0      0  
3             0     0    0     0      0  
4             0     0    0     0      0  
...         ...   ...  ...   ...    ...  
2325          0     0    0     0      0  
2326          0     0    0     1      0  
2327          0     0    0     0      0  
2328          0     0    0     0      0  
2329          0     0    0     0      0  

[2330 rows x 163 columns]

### Random Forests

In [36]:
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, test_size=0.2, random_state=42)

In [43]:
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)
print("F1-Score:", str(f1_score(y_test, y_pred_rf, average = 'micro')))
print(classification_report(y_test, y_pred_rf, zero_division=0))

F1-Score: 0.8606741573033708
              precision    recall  f1-score   support

           0       0.79      0.61      0.69        38
           1       1.00      1.00      1.00        76
           2       1.00      0.96      0.98        50
           3       1.00      0.94      0.97        16
           4       0.97      0.91      0.94        35
           5       0.77      0.72      0.74        46
           6       0.96      0.89      0.92        53
           7       1.00      0.44      0.61        32
           8       0.90      0.82      0.86        98
           9       0.65      0.56      0.60        27

   micro avg       0.91      0.81      0.86       471
   macro avg       0.90      0.78      0.83       471
weighted avg       0.91      0.81      0.85       471
 samples avg       0.82      0.82      0.82       471



Random Forest Optimization

In [38]:
rf_model = RandomForestClassifier(random_state=42)
param_grid = {
    'n_estimators': list(range(20,300,50)), 
    'max_depth': list(range(3,60,3)), 
    'min_samples_split': [2, 5, 10, 20, 30], 
    'min_samples_leaf': [1, 2, 4, 10]
}
scoring = {'F1-Score': make_scorer(f1_score, average='micro', needs_proba=False)}
grid_search = GridSearchCV(rf_model, param_grid, cv=5, scoring=scoring, refit='F1-Score', n_jobs=-1)

grid_search.fit(X_train, y_train)
print('Best parameters found: \n', grid_search.best_params_)

Best parameters found: 
 {'max_depth': 42, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 170}


In [41]:
grid_search.best_estimator_

RandomForestClassifier(max_depth=42, n_estimators=170, random_state=42)

Testing tuned parameters

In [42]:
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, test_size=0.2, random_state=42)
rf_model = RandomForestClassifier(max_depth=42, n_estimators=170, random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)
print("F1-Score:", str(f1_score(y_test, y_pred_rf, average = 'micro')))
print(classification_report(y_test, y_pred_rf, zero_division=0))

F1-Score: 0.867630700778643
              precision    recall  f1-score   support

           0       0.83      0.63      0.72        38
           1       1.00      0.99      0.99        76
           2       1.00      0.96      0.98        50
           3       1.00      0.94      0.97        16
           4       0.97      0.91      0.94        35
           5       0.70      0.65      0.67        46
           6       0.96      0.89      0.92        53
           7       1.00      0.50      0.67        32
           8       0.90      0.89      0.89        98
           9       0.70      0.59      0.64        27

   micro avg       0.91      0.83      0.87       471
   macro avg       0.90      0.79      0.84       471
weighted avg       0.91      0.83      0.86       471
 samples avg       0.83      0.83      0.83       471



#### Support Vector Machines

In [44]:
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, test_size=0.2, random_state=42)

In [45]:
svm_model = OneVsRestClassifier(SVC(random_state=42))
svm_model.fit(X_train, y_train)
y_pred_svm = svm_model.predict(X_test)
print(f1_score(y_test, y_pred_svm, average='micro'))
print(classification_report(y_test, y_pred_svm, zero_division=0))

0.8429378531073446
              precision    recall  f1-score   support

           0       0.92      0.58      0.71        38
           1       1.00      0.97      0.99        76
           2       1.00      0.92      0.96        50
           3       0.78      0.88      0.82        16
           4       0.97      0.80      0.88        35
           5       0.64      0.59      0.61        46
           6       0.96      0.91      0.93        53
           7       1.00      0.44      0.61        32
           8       0.86      0.89      0.87        98
           9       0.81      0.48      0.60        27

   micro avg       0.90      0.79      0.84       471
   macro avg       0.89      0.74      0.80       471
weighted avg       0.90      0.79      0.83       471
 samples avg       0.80      0.79      0.79       471



SVM Optimization

In [46]:
svm_model = OneVsRestClassifier(SVC(random_state=42))

param_grid = {
    'estimator__C': [0.1, 1, 10, 0.01, 0.5, 5, 50],
    'estimator__kernel': ['linear', 'rbf', 'poly'],
    'estimator__gamma': ['scale', 'auto',0.001, 0.01, 0.1, 1, 10]
}

scoring = {'F1-Score': make_scorer(f1_score, average='micro')}
grid_search = GridSearchCV(svm_model, param_grid, scoring=scoring, refit='F1-Score', cv=5, n_jobs=-1)

# Perform the search
grid_search.fit(X_train, y_train)

# Best parameter set
print('Best parameters found: \n', grid_search.best_params_)

In [ ]:
grid_search.best_params_

In [ ]:
grid_search.best_estimator_

Testing tuned parameters

In [81]:
svm_model = OneVsRestClassifier(SVC(C= 5, gamma = 'scale', random_state=42, class_weight='balanced'))
svm_model.fit(X_train, y_train)
y_pred_svm = svm_model.predict(X_test)
print(f1_score(y_test, y_pred_svm, average='micro'))
print(classification_report(y_test, y_pred_svm, zero_division=0))


0.8355739400206825
              precision    recall  f1-score   support

           0       0.52      0.76      0.62        38
           1       1.00      0.99      0.99        76
           2       0.98      0.94      0.96        50
           3       0.71      0.94      0.81        16
           4       0.92      0.97      0.94        35
           5       0.62      0.72      0.67        46
           6       0.93      0.94      0.93        53
           7       1.00      0.53      0.69        32
           8       0.88      0.88      0.88        98
           9       0.49      0.67      0.56        27

   micro avg       0.81      0.86      0.84       471
   macro avg       0.80      0.83      0.81       471
weighted avg       0.84      0.86      0.84       471
 samples avg       0.80      0.86      0.82       471

